components/script/script_module.rs

```diff
@@ -697,158 +834,64 @@ impl ModuleOwner {
         }
     }
 
-    fn gen_promise_with_final_handler(
-        &self,
-        module_url: Option<ServoUrl>,
-        is_top_level: bool,
-    ) -> Rc<Promise> {
-        let resolve_this = self.clone();
-        let reject_this = self.clone();
-
-        let resolved_url = module_url.clone();
-        let rejected_url = module_url.clone();
-
-        let handler = PromiseNativeHandler::new(
-            &self.global(),
-            Some(ModuleHandler::new(Box::new(
-                task!(fetched_resolve: move || {
-                    resolve_this.finish_module_load(resolved_url, is_top_level);
-                }),
-            ))),
-            Some(ModuleHandler::new(Box::new(
-                task!(failure_reject: move || {
-                    reject_this.finish_module_load(rejected_url, is_top_level);
-                }),
-            ))),
-        );
-
-        let realm = enter_realm(&*self.global());
-        let comp = InRealm::Entered(&realm);
-        let _ais = AutoIncumbentScript::new(&*self.global());
-
-        let promise = Promise::new_in_current_realm(&self.global(), comp);
-
-        promise.append_native_handler(&handler);
-
-        promise
-    }
-
-    /// https://html.spec.whatwg.org/multipage/#fetch-the-descendants-of-and-link-a-module-script
-    /// step 4-7.
-    pub fn finish_module_load(&self, module_url: Option<ServoUrl>, is_top_level: bool) {
+    pub fn notify_owner_to_finish(&self, module_identity: ModuleIdentity) {
         match &self {
             ModuleOwner::Worker(_) => unimplemented!(),
             ModuleOwner::Window(script) => {
                 let global = self.global();
 
                 let document = document_from_node(&*script.root());
 
-                let module_map = global.get_module_map().borrow();
-
-                let (module_tree, mut load) = if let Some(script_src) = module_url.clone() {
-                    let module_tree = module_map.get(&script_src.clone()).unwrap().clone();
-
-                    let load = Ok(ScriptOrigin::external(
-                        module_tree.get_text().borrow().clone(),
-                        script_src.clone(),
-                        ScriptType::Module,
-                    ));
-
-                    debug!(
-                        "Going to finish external script from {}",
-                        script_src.clone()
-                    );
-
-                    (module_tree, load)
-                } else {
-                    let module_tree = {
-                        let inline_module_map = global.get_inline_module_map().borrow();
-                        inline_module_map
-                            .get(&script.root().get_script_id())
-                            .unwrap()
-                            .clone()
+                let load = {
+                    let module_tree = match module_identity.clone() {
+                        ModuleIdentity::ModuleUrl(script_src) => {
```

Comment:
It seems like "get the module tree for an identity" is an operation that's performed often, factor it out into a method?